gcd/src/main/java/org/spine3/server/storage/datastore/Indexes.java

```diff
@@ -0,0 +1,108 @@
+ * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+ * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+package org.spine3.server.storage.datastore;
+
+import com.google.cloud.datastore.Entity;
+import com.google.cloud.datastore.EntityQuery;
+import com.google.cloud.datastore.Key;
+import com.google.cloud.datastore.Query;
+import com.google.cloud.datastore.StructuredQuery;
+import com.google.common.base.Function;
+import com.google.common.collect.Iterators;
+import com.google.common.reflect.TypeToken;
+import org.spine3.base.Stringifiers;
+
+import javax.annotation.Nullable;
+import java.util.Iterator;
+
+import static com.google.common.base.Preconditions.checkNotNull;
+
+/**
+ * A utility for generating the {@linkplain org.spine3.server.storage.Storage#index() storage ID indexes}.
+ *
+ * @author Dmytro Dashenkov.
+ * @see org.spine3.server.storage.Storage#index()
+ */
+public class Indexes {
+
+    private Indexes() {
+    }
+
+    /**
+     * Retrieves the ID index for the given {@code kind} or the storage records.
+     *
+     * <p>A call to this method is equivalent to {@code indexIterator(datastore, kind, idClass, null)}.
+     *
+     * @param datastore datastore to get the indexes for
+     * @param kind      the kind if the records in the datastore
+     * @param idClass   {@link Class} of the record ID
+     * @param <I>       type of the IDs to retrieve
+     * @return an {@link Iterator} of the IDs matching given record kind
+     * @see #indexIterator(DatastoreWrapper, String, Class, StructuredQuery.Filter)
+     */
+    public static <I> Iterator<I> indexIterator(DatastoreWrapper datastore,
+                                                String kind,
+                                                Class<I> idClass) {
+        return indexIterator(datastore, kind, idClass, null);
+    }
+
+    /**
+     * Retrieves the ID index for the given {@code kind} or the storage records.
+     *
+     * @param datastore datastore to get the indexes for
+     * @param kind      the kind if the records in the datastore
+     * @param idClass   {@link Class} of the record ID
+     * @param filter    custom filters for selecting only a subset of the records of given type from datastore;
+     *                  {@code null} stands for no custom filters
+     * @param <I>       type of the IDs to retrieve
+     * @return an {@link Iterator} of the IDs matching given record kind and filter
+     */
+    public static <I> Iterator<I> indexIterator(DatastoreWrapper datastore,
+                                                String kind,
+                                                Class<I> idClass,
+                                                @Nullable StructuredQuery.Filter filter) {
+        checkNotNull(datastore, "datastore");
```

Comment:
After some further thought, I think we can do without the param names. The information on which param is `null` would be available by stack trace of a through exception. Since we have each check on its own line, we'd be able to find the cause.